File: scripts/generate-favicon.js
import sharp from 'sharp';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const logoPath = path.join(__dirname, '../public/images/cardlogo.png');
const outputDir = path.join(__dirname, '../public');

// Standard favicon sizes
const sizes = [
  { size: 16, name: 'favicon-16x16.png' },
  { size: 32, name: 'favicon-32x32.png' },
  { size: 96, name: 'favicon-96x96.png' },
  { size: 192, name: 'android-chrome-192x192.png' },
  { size: 512, name: 'android-chrome-512x512.png' },
  { size: 180, name: 'apple-touch-icon.png' }, // Apple touch icon
];

async function generateFavicons() {
  try {
    console.log('Generating favicons from logo...');

    // Check if logo exists
    if (!fs.existsSync(logoPath)) {
      console.error(`Logo not found at ${logoPath}`);
      process.exit(1);
    }

    // Generate each size
    for (const { size, name } of sizes) {
      const outputPath = path.join(outputDir, name);

      await sharp(logoPath)
        .resize(size, size, {
          fit: 'contain',
          background: { r: 255, g: 255, b: 255, alpha: 0 }
        })
        .png()
        .toFile(outputPath);

      console.log(`✓ Generated ${name} (${size}x${size})`);
    }

    // Generate ICO file (16x16 and 32x32 combined)
    const ico16 = await sharp(logoPath)
      .resize(16, 16, {
        fit: 'contain',
        background: { r: 255, g: 255, b: 255, alpha: 0 }
      })
      .png()
      .toBuffer();

    const ico32 = await sharp(logoPath)
      .resize(32, 32, {
        fit: 'contain',
        background: { r: 255, g: 255, b: 255, alpha: 0 }
      })
      .png()
      .toBuffer();

    // For ICO, we'll just copy the 32x32 as favicon.ico (simplified approach)
    // Most modern browsers accept PNG as favicon.ico
    fs.writeFileSync(
      path.join(outputDir, 'favicon.ico'),
      ico32
    );
    console.log('✓ Generated favicon.ico');

    // Create site.webmanifest for Android
    const manifest = {
      name: 'CARD',
      short_name: 'CARD',
      icons: [
        {
          src: '/android-chrome-192x192.png',
          sizes: '192x192',
          type: 'image/png'
        },
        {
          src: '/android-chrome-512x512.png',
          sizes: '512x512',
          type: 'image/png'
        }
      ],
      theme_color: '#ffffff',
      background_color: '#ffffff',
      display: 'standalone'
    };

    fs.writeFileSync(
      path.join(outputDir, 'site.webmanifest'),
      JSON.stringify(manifest, null, 2)
    );
    console.log('✓ Generated site.webmanifest');

    console.log('\n✅ All favicons generated successfully!');
  } catch (error) {
    console.error('Error generating favicons:', error);
    process.exit(1);
  }
}

generateFavicons();

File: scripts/generate-favicon.py
#!/usr/bin/env python3
import sys
import json
from pathlib import Path

try:
    from PIL import Image
except ImportError:
    print("Error: PIL (Pillow) not installed. Install with: pip install Pillow")
    sys.exit(1)

LOGO_PATH = Path("public/images/cardlogo.png")
OUTPUT_DIR = Path("public")

# Check if logo exists
if not LOGO_PATH.exists():
    print(f"Error: Logo not found at {LOGO_PATH}")
    sys.exit(1)

print("Generating favicons from logo...")

# Open the logo
logo = Image.open(LOGO_PATH)

# Standard favicon sizes
sizes = [
    (16, "favicon-16x16.png"),
    (32, "favicon-32x32.png"),
    (96, "favicon-96x96.png"),
    (192, "android-chrome-192x192.png"),
    (512, "android-chrome-512x512.png"),
    (180, "apple-touch-icon.png"),
]

# Generate each size
for size, filename in sizes:
    # Resize maintaining aspect ratio, then create square with transparent background
    resized = logo.resize((size, size), Image.Resampling.LANCZOS)
    output_path = OUTPUT_DIR / filename
    resized.save(output_path, "PNG")
    print(f"✓ Generated {filename} ({size}x{size})")

# Generate ICO file (using 32x32)
ico = logo.resize((32, 32), Image.Resampling.LANCZOS)
ico.save(OUTPUT_DIR / "favicon.ico", "ICO")
print("✓ Generated favicon.ico")

# Create site.webmanifest
manifest = {
    "name": "CARD",
    "short_name": "CARD",
    "icons": [
        {
            "src": "/android-chrome-192x192.png",
            "sizes": "192x192",
            "type": "image/png"
        },
        {
            "src": "/android-chrome-512x512.png",
            "sizes": "512x512",
            "type": "image/png"
        }
    ],
    "theme_color": "#ffffff",
    "background_color": "#ffffff",
    "display": "standalone"
}

manifest_path = OUTPUT_DIR / "site.webmanifest"
with open(manifest_path, "w") as f:
    json.dump(manifest, f, indent=2)
print("✓ Generated site.webmanifest")

print("\n✅ All favicons generated successfully!")
